feat: add getByUserName endpoint for user login and update tests

HijoDeHonor opened the pull request:
- Implemented GET endpoint '/api/users/' for user login - Added unit tests for the new endpoint.

HijoDeHonor committed:
feat: add getByUserName endpoint for user login and update tests

  - Implemented GET endpoint '/api/users/' for user login
    - Added unit tests for the new endpoint
    - Added semicolons for code consistency

PiewDev commented on errors/ErrorTypes/userDontExist.js:
```diff
@@ -0,0 +1,9 @@
+import { USER_DONT_EXIST_ERROR } from "../../utils/textConstants.js";
+
+export class UserDontExistError extends Error {
```
The error name is incorrectly conjugated.(is: doesn't) Maybe DoesNotExistError(entity) it's more generic implementation

PiewDev commented on errors/errorHandler/errorhandler.js:
```diff
@@ -1,12 +1,14 @@
-import { ALREADY_EXIST_ERROR, INTERNAL_SERVER_ERROR, INVALID_DATA_ERROR } from "../../utils/textConstants.js"
+import { ALREADY_EXIST_ERROR, INTERNAL_SERVER_ERROR, INVALID_DATA_ERROR, INVALID_LOGIN_ERROR, USER_DONT_EXIST_ERROR } from "../../utils/textConstants.js";
 
```
It will separate the error dictionary to another file, more than anything to avoid editing the handler file unnecessarily.

PiewDev commented on test/user/userControler.test.js:
```diff
@@ -59,27 +59,108 @@ describe('create', () => {
   });
 
 });
+
```
Separate the tests by use case: 
RegisterUserController.Test 
LoginUserController.Test

PiewDev commented on test/user/userModel.test.js:
```diff
@@ -37,28 +40,86 @@ describe('UserModel', () => {
+    expect(executeQueryMock).toHaveBeenCalledTimes(2);
+  });
+});
+
```
Separate the tests by use case:
RegisterUserModel.Test
LoginUserModel.Test

PiewDev commented on users/registerUser/userControler.js:
```diff
@@ -17,5 +18,27 @@ export class UserController {
+  });
+  getByUserName = tryCatch(async (req, res) => {
+    const { userName, password } = req.body;
+    const result = validateUser(userName, password);
```
"result" is not well understood, perhaps a better name is "validation".

PiewDev commented on users/registerUser/userControler.js:
```diff
@@ -17,5 +18,27 @@ export class UserController {
+
+    const logUser = await this.userModel.getByUserName({ input: result.data });
+    if (logUser) {
+      const { token, ...userWithoutToken } = logUser;
```
deconstructing here is not good practice.

PiewDev commented on users/userModel.js:
```diff
@@ -1,35 +1,70 @@
+    }
+  }
+
+  static async getByUserName ({ input }) {
```
Maybe I would make this more generic: getBy(property, value);
UserModel.getBy('userName', 'admin');

PiewDev commented on users/userModel.js:
```diff
@@ -1,35 +1,70 @@
+        [userName]
+      );
+
+      if (rows.length === 0) throw new UserDontExistError(USER_DONT_EXIST, USERS);
```
when you do an if you should always do a break with {}

if (rows.length === 0) {
    throw new UserDontExistError(USER_DONT_EXIST, USERS);
}

PiewDev commented on users/userModel.js:
```diff
@@ -1,35 +1,70 @@
+      if (user.password !== password) {
+        throw new InvalidLoginError(INVALID_LOGIN, USERS);
+      }
+      const token = jwt.sign({ userName: user.userName },
```
handling of the jwt token is not the responsibility of the model. Move this logic to other file and call them from controller.

PiewDev reviewed (changes requested):
There are some separation of responsibilities issues to resolve and some improvements that could be good to implement. I recommend using a Linter like EsLint to avoid PRs with many code style changes.

HijoDeHonor committed:
refactor: reorganize tests and repositories, clarify responsibilities

- Separated tests by use case
- Renamed model to repository and clarified its responsibilities
- Moved JWT creation to a separate service file
- Created an error dictionary file
- Fixed typos and spelling errors

PiewDev commented on errors/ErrorTypes/userDoesNotExist.js:
```diff
@@ -0,0 +1,9 @@
+import { USER_DOES_NOT_EXIST_ERROR } from "../../utils/textConstants.js";
+
+export class UserDoesNotExistError extends Error {
```
You are using entity to make the error generic shouldnt be UserDoestNotExistError but DoestNotExistError

HijoDeHonor committed:
fix: correct test implementation

Fixed an error in the way a test was implemented to ensure accurate and reliable test results.

PiewDev commented on errors/ErrorTypes/userAlreadyExist.js:
The file has not the correct name

PiewDev commented on users/mockModel.js:
Delete this file

PiewDev commented on users/registerUser/userControler.js:
```diff
@@ -1,21 +1,51 @@
   create = tryCatch(async (req, res) => {
-    const { userName, password } = req.body
+    const { userName, password } = req.body;
     const result = validateUser(userName, password);
```
change result to validation too

PiewDev commented on users/userRepository.js:
methods shouldnt be static

PiewDev commented on users/userRepository.js:
```diff
@@ -0,0 +1,56 @@
+    }
+  }
+
+  static async getBy (property, value) {
```
here should be: 
`getBy (property) {
const key=Object.keys(property)[0]
const value=Object.values(property)[0]
}
`

PiewDev replied:
This method should be called: 
`model.getBy({userName = 'admin'})`

PiewDev reviewed (changes requested):
I cleared up some confusion and found some small improvements to do.

HijoDeHonor committed:
refactor: update file names and methods, add and abstract error types

- Corrected names in various files for better clarity and consistency
- Changed methods to non-static
- Added new error types and abstracted some existing errors for better error handling

PiewDev reviewed (approved):
Good job!
